rustls/src/client/ech.rs

```diff
@@ -0,0 +1,509 @@
+    pub(crate) fn transcript_hrr_update(&mut self, hash: &'static dyn Hash, m: &Message) {
+        trace!("Updating ECH inner transcript for HRR");
+
+        let inner_transcript = self
+            .inner_hello_transcript
+            .get_fork()
+            .start_hash(hash);
+
+        let mut inner_transcript_buffer = inner_transcript.into_hrr_buffer();
+        inner_transcript_buffer.add_message(m);
+        self.inner_hello_transcript = inner_transcript_buffer;
+    }
+
+    // Confirm whether an ECH offer was accepted based on examining the server hello.
+    pub(crate) fn confirm_acceptance(
+        self,
+        ks: &mut KeyScheduleHandshakeStart,
+        server_hello: &ServerHelloPayload,
+        hash: &'static dyn Hash,
+    ) -> Result<Option<(HandshakeHash, Random)>, Error> {
+        // Star the inner transcript hash now that we know the hash algorithm to use.
+        let inner_transcript = self
+            .inner_hello_transcript
+            .start_hash(hash);
+
+        // Fork the transcript that we've started with the inner hello to use for a confirmation step.
+        // We need to preserve the original inner_transcript to use if this confirmation succeeds.
+        let mut confirmation_transcript = inner_transcript.get_fork();
+
+        // Add the server hello confirmation - this differs from the standard server hello encoding.
+        confirmation_transcript.add_message(&Self::server_hello_conf(server_hello));
+
+        // Derive a confirmation secret from the inner hello random and the confirmation transcript.
+        let derived = ks.server_ech_confirmation_secret(
+            self.inner_hello_random.0.as_ref(),
+            confirmation_transcript.get_current_hash(),
+        );
+
+        // Check that first 8 digits of the derived secret match the last 8 digits of the original
+        // server random. This match signals that the server accepted the ECH offer.
+        // Indexing safety: Random is [0; 32] by construction.
+        match ConstantTimeEq::ct_eq(derived.as_ref(), server_hello.random.0[24..].as_ref()).into() {
+            true => {
+                trace!("ECH accepted by server");
+                Ok(Some((inner_transcript, self.inner_hello_random)))
+            }
+            false => {
+                trace!("ECH rejected by server");
+                Ok(None)
+            }
+        }
+    }
+
+    pub(crate) fn confirm_hrr_acceptance(
+        &self,
+        hrr: &HelloRetryRequest,
+        cs: &Tls13CipherSuite,
+        common: &mut CommonState,
+    ) -> Result<(), Error> {
+        // The client checks for the "encrypted_client_hello" extension.
+        let ech_conf = match hrr.get_ech_retry_request() {
+            // If none is found, the server has rejected ECH and there are no retry
+            // configs to offer. Abort with an ech required alert and return an err.
+            None => {
+                return Err(fatal_alert_required(common, None));
```

Comment:
Style nit: this can sometimes be more compact if you omit the `;` (because it doesn't need a block).

Reply:
Good call, done.